### PR TITLE
Upgrade status-go to 0.17.0 (Constantinople).

### LINKED ISSUE #1258: [ULC/LES] Initial sync never finishes
# Problem

I can't make ULC to finish the initial sync properly. It ends up with these log entries repeating.

Syncing from scratch, the datadir/keydir are cleared before starting.

```
WARN [10-26|09:46:33.299] Rolled back headers                      count=112  header=4229887->4229887 fast=0->0 block=0->0
DEBUG[10-26|09:46:33.300] Transaction receipt download terminated  err="receipt download canceled (requested)"
DEBUG[10-26|09:46:33.300] Skeleton fill terminated                 err="block header download canceled (requested)"
DEBUG[10-26|09:46:33.300] Skeleton chain invalid                   peer=690c93ac2f6e6f71 err="block header download canceled (requested)"
DEBUG[10-26|09:46:33.300] Header download terminated               peer=690c93ac2f6e6f71
DEBUG[10-26|09:46:33.300] Block body download terminated           err="block body download canceled (requested)"
DEBUG[10-26|09:46:33.300] Synchronisation terminated               elapsed=384.666263ms
WARN [10-26|09:46:33.300] Synchronisation failed, dropping peer    peer=690c93ac2f6e6f71 err="retrieved hash chain is invalid"
DEBUG[10-26|09:46:33.300] Done synchronising with peer             id=690c93ac2f6e6f71 conn=staticdial
DEBUG[10-26|09:46:33.300] Unknown peer to check sync headers       id=690c93ac2f6e6f71 conn=staticdial
DEBUG[10-26|09:46:33.300] Light Ethereum message handling failed   id=690c93ac2f6e6f71 conn=staticdial err=EOF
DEBUG[10-26|09:46:33.300] Disconnected old entry                   enode=690c93ac2f6e6f71
DEBUG[10-26|09:46:33.300] Removing p2p peer                        id=690c93ac2f6e6f71 conn=staticdial duration=15.243s   peers=2 req=false err="useless peer"
```

At the same time, `eth_syncing` is showing that the sync is happening:
```
{
	"jsonrpc": "2.0",
	"id": 121212,
	"result": {
		"currentBlock": "0x408aff",
		"highestBlock": "0x41ab35",
		"knownStates": "0x0",
		"pulledStates": "0x0",
		"startingBlock": "0x3dffff"
	}
}
```
statusd, build from sources @ `bdbceba6eb3c1065c1aca235b13791f1e95fe739`

statusd config: [ulc-conf.json.zip](https://github.com/status-im/status-go/files/2518245/ulc-conf.json.zip)

@b00ris @JekaMas can you take a look?

### PR DESCRIPTION
**Pull request by mandrigin**
Upgrading status-go to the Constantinople-supporting version (required for LES/ULC to work correctly).

fixes: https://github.com/status-im/status-go/issues/1258

status: ready

### CODE REVIEW

**Review by pedropombeiro — APPROVED** [review]

**Comment by Serhy** [comment]
There is no offline message history is fetched with this status go in public / 1-1 chats.
[geth.log](https://github.com/status-im/status-react/files/2615371/geth.7.log)

[logcat.txt](https://github.com/status-im/status-react/files/2615436/logcatStEthByPid.txt) after app start with public chat awaiting history

**Comment by mandrigin** [comment]
@Serhy does the blockchain sync on LES?

**Comment by Serhy** [comment]
It behaved differently. Sync was happening (but not more than 4230000 block) but now it doesn't start sync at all: 
[geth.log](https://github.com/status-im/status-react/files/2616530/geth.8.log) for Ropsten (LES) network.

**Comment by mandrigin** [comment]
@Serhy thanks! I will take a look more closely.

**Comment by mandrigin** [comment]
hmm, strange, it syncs when I run statusgo from the same tag as 0.17.0 in command-line...

**Comment by Serhy** [comment]
@mandrigin Yes, and from n-th app restart I can confirm on my side sync finished (with `LES: 'latest' 4329035` by that moment on iOS)
It appears that in some non-rare cases sync fails. And just now, after sync on my side suddenly stopped somewhere in the middle (to be precise it was stuck on `syncing 4284415 of 4330522 blocks...`) I restarted the app -> sync did not start after I waited 5+ min and tail geth.log shows the same as what was before in https://github.com/status-im/status-react/pull/6877#issuecomment-441711053:


```t=2018-11-26T18:28:36+0000 lvl=info msg="Initialized encryption service"               package=status-go/services/sshext.chat installationID=5208492f-7c3a-5936-9198-9c6c15a01f5c
t=2018-11-26T18:28:47+0000 lvl=info msg="message loop"                                 peer=8f7e7c599f831d153f5a8ef24a1c5333f4e3320cfa0474be2f64cacc11d29e68 err=EOF
t=2018-11-26T18:29:02+0000 lvl=eror msg="Failed to register sync peer"                 peer=0419bab5f2fb5f64                                                 err="peer is already registered"
t=2018-11-26T18:29:10+0000 lvl=eror msg="Failed to register sync peer"                 peer=5eb5c2ac92c8ae35                                                 err="peer is already registered"
t=2018-11-26T18:29:19+0000 lvl=info msg="message loop"                                 peer=8f7e7c599f831d153f5a8ef24a1c5333f4e3320cfa0474be2f64cacc11d29e68 err=EOF
t=2018-11-26T18:29:32+0000 lvl=eror msg="Failed to register sync peer"                 peer=f9f53ce1ff778dcd                                                 err="peer is already registered"
t=2018-11-26T18:29:32+0000 lvl=eror msg="Failed to register sync peer"                 peer=0419bab5f2fb5f64                                                 err="peer is already registered"
t=2018-11-26T18:29:43+0000 lvl=eror msg="Failed to register sync peer"                 peer=5eb5c2ac92c8ae35                                                 err="peer is already registered"
t=2018-11-26T18:30:02+0000 lvl=eror msg="Failed to register sync peer"                 peer=f9f53ce1ff778dcd                                                 err="peer is already registered"
t=2018-11-26T18:30:38+0000 lvl=info msg="App State changed"                            package=status-go/api.StatusBackend    new-state=background

```

So sync with upgraded status-go to 0.17.0 'Ropsten LES' now supports Constantinople.

However sync may not start in some cases of app re-start, which is confirmed by geth.log's `msg="Failed to register sync peer" err="peer is already registered"` (if i'm correct that it is a reason of not syncing...)

**Review by rasom — APPROVED** [review]

**Review by dshulyak — CHANGES_REQUESTED** [review]
just so that it won't get merged by accident. this change will make requesting history broken.

**Comment by mandrigin** [comment]
@dshulyak thanks!

**Comment by mandrigin** [comment]
@adambabik do you have any thoughts on why it can't receive messages from the mailservers?

**Comment by adambabik** [comment]
I have no idea really. I use `0.17.0` in bots and they work fine. I run this build on iOS device and I can't really tell anything: whether it connected to any mail server, whether it sent any requests, nothing. I am blind 🙈

**Comment by adambabik** [comment]
@Serhy can you try to run the app again and get more logs, please? The one you posted (`geth.log`) is quite empty.

```
t=2018-11-26T11:54:53+0000 lvl=info msg="Initialized encryption service" package=status-go/services/sshext.chat installationID=645d11a0-a3b7-56f1-b61e-ac8030b9aa73
t=2018-11-26T11:55:16+0000 lvl=info msg="App State changed"              package=status-go/api.StatusBackend    new-state=background
```

As you can see there is a gap of 23 seconds and it looks like nothing happened. Also, please run it with `DEBUG` log level.

**Comment by Serhy** [comment]
@adambabik sure, yes, here is the geth.log with DEBUG log level for the app session (in the scope of that session contact joined public chat where history supposed to be fetched and displayed in chat):
[gethDbgLevelmailservers.log](https://github.com/status-im/status-react/files/2621202/gethDbgLevelmailservers.log)

**Comment by adambabik** [comment]
@Serhy thanks! It looks fine. I don't see any request being sent by the app and we have a log that should be visible: https://github.com/status-im/status-go/blob/0.17.0/services/shhext/api.go#L155. Can someone from status-react help debug the client?

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/status-im/status-go">status-im/status-go</a></strong></div><div>The Status module that consumes go-ethereum. Contribute to status-im/status-go development by creating an account on GitHub.</div></blockquote>

**Comment by mandrigin** [comment]
@adambabik I can take a look at the react side later today, because it seems to block upgrades...

**Comment by dshulyak** [comment]
Just spotted that there was one extension for information that was returned from discovery.summary. Now each peer also has enode field, like:

```json
{
    "enode": "enode://76ee16566fb45ca7644c8dec7ac74cadba3bfa0b92c566ad07bcb73298b0ffe1315fd787e1f829e90dba5cd3f4e0916e069f14e50e9cbec148bead397ac8122d@47.91.226.75:30305",
    "id": "84095cf2c0c727fc244fba8feba5da07e82a8cdd4d7fe406c31443e0f1013510",
    "name": "Statusd/v0.16.8/linux-amd64/go1.10.2",
    "caps": [
      "shh/6"
    ],
    "network": {
      "localAddress": "192.168.0.107:55002",
      "remoteAddress": "47.91.226.75:30305",
      "inbound": false,
      "trusted": false,
      "static": true
    },
    "protocols": {
      "shh": "unknown"
    }
  }
```

could it be an issue?

**Comment by dshulyak** [comment]
Actually id field is marshaled differently. I would like someone to confirm, but i think that status react waits for a peer with specific ID to appear in discovery.summary. Is it the case?

**Comment by mandrigin** [comment]
@dshulyak let me check.

**Comment by dshulyak** [comment]
Looks like this is it. I created separate PR to verify me guess https://github.com/status-im/status-react/pull/6919

**Comment by mandrigin** [comment]
@dshulyak it looks like you are right, this method always returns `false` now.

See the example parameters.
**enode**
```
enode://da61e9eff86a56633b635f887d8b91e0ff5236bbc05b8169834292e92afb92929dcf6efdbf373a37903da8fe0384d5a0a8247e83f1ce211aa429200b6d28c548@47.91.156.93:30504 
```
**peers**
```
[
  {:enode "enode://887cbd92d95afc2c5f1e227356314a53d3d18855880ac0509e0c0870362aee03939d4074e6ad31365915af41d34320b5094bfcc12a67c381788cd7298d06c875@206.189.243.177:30305?discport=0", :id "410f5dcad03a883948d31aaa04887df5c5281d1473751ef5a89396cdcd2751e2", :name "Statusd/v0.16.8/linux-amd64/go1.10.2", :caps ["shh/6"], :network {:localAddress "192.168.0.9:50070", :remoteAddress "206.189.243.177:30305", :inbound false, :trusted false, :static true}, :protocols {:shh "unknown"}}
  {:enode "enode://da61e9eff86a56633b635f887d8b91e0ff5236bbc05b8169834292e92afb92929dcf6efdbf373a37903da8fe0384d5a0a8247e83f1ce211aa429200b6d28c548@47.91.156.93:30504", :id "82c196d7a77d3a73d7a3b0bb7ec43ed8d6b758703815c4b7191332170cadd331", :name "Statusd/v0.16.8/linux-amd64/go1.10.2", :caps ["shh/6"], :network {:localAddress "192.168.0.9:50053", :remoteAddress "47.91.156.93:30504", :inbound false, :trusted false, :static true}, :protocols {:shh "unknown"}}
  {:enode "enode://7c80e37f324bbc767d890e6381854ef9985d33940285413311e8b5927bf47702afa40cd5d34be9aa6183ac467009b9545e24b0d0bc54ef2b773547bb8c274192@47.91.155.62:30305", :id "84d828cecc6d7fa6e4f831501e528f8593999d9972d4e5b46a22defb5fe554b2", :name "Statusd/v0.16.8/linux-amd64/go1.10.2", :caps ["shh/6"], :network {:localAddress "192.168.0.9:49985", :remoteAddress "47.91.155.62:30305", :inbound false, :trusted false, :static true}, :protocols {:shh "unknown"}}
  {:enode "enode://03f21684e777341df7f1b93888c9ba13a4d15130bb3c2d4fb13e817035b6cd6f3ed3dad1cd6c48dc995b17a2f4005188bae6d74d19d361a789c0f7085675a7f2@206.189.243.177:30304?discport=0", :id "9542a8e719ff50e4bcb364bc50c0e91e7bf1c91d4bac26bb48097e021f032876", :name "Statusd/v0.16.8/linux-amd64/go1.10.2", :caps ["shh/6"], :network {:localAddress "192.168.0.9:50068", :remoteAddress "206.189.243.177:30304", :inbound false, :trusted false, :static true}, :protocols {:shh "unknown"}}
  {:enode "enode://aa97756bc147d74be6d07adfc465266e17756339d3d18591f4be9d1b2e80b86baf314aed79adbe8142bcb42bc7bc40e83ee3bbd0b82548e595bf855d548906a1@47.52.188.241:30305", :id "a681612c0a4b3321926dfe00af7bb5891d07a512a24b5ba6b05e11703d66aec3", :name "Statusd/v0.16.8/linux-amd64/go1.10.2", :caps ["shh/6"], :network {:localAddress "192.168.0.9:49986", :remoteAddress "47.52.188.241:30305", :inbound false, :trusted false, :static true}, :protocols {:shh "unknown"}}
]
```

so with the current code this will never be found

**Comment by mandrigin** [comment]
but even with your change I can't seem to make it request messages, I'll debug a bit more after lunch

**Comment by mandrigin** [comment]
@dshulyak okay, it was my bad, it looks like your change helps. I will push the fix here.

**Review by dshulyak — APPROVED** [review]

**Comment by mandrigin** [comment]
@Serhy so I hope I fixed mailserver requests, but please make sure to test everything very thoroughly, because of the change in the internal datastructures of go-ethereum, there might be more sneaky bugs ☹️

PS: issues with LES enabled aren't blockers for this issue, they are just to be registered.

**Comment by mandrigin** [comment]
unit tests 👿

**Comment by mandrigin** [comment]
now they should pass

**Comment by antdanchenko** [comment]
I am going to re-run e2e tests due to internal error during first run

**Comment by cammellos** [comment]
@mandrigin could you upgrade to 0.17.1 please? @adambabik prepared a release, and should only contain non-breaking changes on top. Thanks!

**Comment by mandrigin** [comment]
@Serhy I've upgrades status-go here and rebased the branch. Please, take a look when you have time.

**Comment by cammellos** [comment]
Once this has been merged I will need to upgrade once more status-go to port back the hotfix that we applied to release branch, before we release desktop. This will be done in a separate PR.

**Comment by churik** [comment]
Tested (On Mac OsX):
- account creation
- recovering account
- fetching history for 24 h
- messaging in 1-1, public, group chats
- upgrading from current nightly
- mailserver switching

@cammellos noticed issue with syncing messages in group chats (my messages from release build 0.9.32 with hotfix are not shown in group chats, if I have sync enabled)

### Steps:
- on mobile have 0.9.32 with hotfix (using account A) or current PR build
- on desktop upgraded from current nightly to PR Build, recovered account A
- enable sync
- all info was synced sucessfully
- create new group chat
- send some messages from desktop -> can see on mobile
- send some messages from mobile -> can't see on desktop

Geth. log from desktop (contains only steps for  issue reproduction):
[geth.log](https://github.com/status-im/status-react/files/2658235/geth.log)

It is not reproducible on current nightly.

**Comment by cammellos** [comment]
Thanks @churik I will be looking into this.

**Comment by Serhy** [comment]
On iOS 12.0.1 and Android 7 tested:
1) transacting in 1-1 chats with ETH and all test tokens Status support withing Ropsten (SNT, ADX, BNT, BAT MANA token transactions in Mainnet)
2) transaction from Wallet
3) transaction in Dapps 
* SimpleTestDapp in Ropsten (all possible activities Dapp has)
* CryptoKitties: Sign In, Buy cat, Breed cats tx
* In the ENS Dapp can register ensname
* AirSwap can Sign, can sign tranasctions to SELL SNT (here transaction is sent but not put into block - same in 0.9.32 - will recheck later again as I don't see it's status-go related (worked 8h earlier with 0.9.32 release build)
* Decentraland can Sign authorising tx and buy MANA

4) Misc. : 
- fleet changes / LogLevel reflects accordingly in geth.log
- web3 permission enabling in SimpleDapp (and CryptoKitties) 
- add/use extention
- messaging in group chat and group chat management
- can switch and sign transaction from Wallet in Rinkeby, POA, xDai chian networks

List of tests could be looking at https://ethstatus.testrail.net/index.php?/runs/view/2127

I want to cover few more Dapps actions...

<blockquote><div><strong><a href="https://ethstatus.testrail.net/index.php?/runs/view/2127">Login - TestRail</a></strong></div></blockquote>

**Comment by cammellos** [comment]
@churik could you test the latest build? 

Basically to replicate the issue:

1) Install release on device A
2) Install nightly on device B
3) Pair the devices
4) Upgrade to PR on device B
5) Exchange a few messages
6) Device B won't be kept in sync

This only happens if you paired the devices for the first time on nightly and upgraded to this PR in less than 6 hours.

After 6 hours if you send a message 1-to-1/group-chat/pairing message from device B the two devices will be again paired correctly. (you can probably manually change the clock to test without waiting).

If you still have the devices that you originally tested with, since more than 6 hours have passed, you can just install the PR and send a message from the device that is not receiving messages, and after that everything should work as expected.

Also this only affect upgrade nightly -> pr (therefore next nightly if merged), 0.9.32 has already the correct version of the code (we are backporting it to develop), so I am inclined to keep as it is, as it's low impact, only those people who have paired their device for the first time on a previous nightly and then within 6 hours have installed the next nightly, which is probably uncommon, and it will fix itself with normal usage.

Basically previous nightly was wrongly advertising the version of a token, while the pr and release correctly do so, but it needs to be refreshed (refreshes every 6 hours) before we can restore normal operations.

I'll try to get a build working but windows is timing out, https://ci.status.im/blue/organizations/jenkins/status-react%2Fpull%20requests/detail/PR-6877/14/pipeline


<blockquote><div><strong><a href="https://ci.status.im/blue/organizations/jenkins/status-react%2Fpull%20requests/detail/PR-6877/14/pipeline">Jenkins</a></strong></div></blockquote>

**Comment by asemiankevich** [comment]
@cammellos i am sorry about this , but would you mind to rebase (if possible ofc). The iOS build is generated as `null`, this one was fixed on Friday i think

**Comment by cammellos** [comment]
@asemiankevich done, waiting for the jenkins lottery :)

**Comment by churik** [comment]
@cammellos issue is not reproducible now on the same dataset. Thanks a lot for finding the reason!
Checked 
1) sync of 1-1, group chats and contacts between devices with PR builds and PR build <-> 0.9.32 with hotfix
2) sync of foto / profile name between devices with PR builds

**Comment by cammellos** [comment]
@churik thanks for spotting the issue!

**Comment by Serhy** [comment]
We've tested few more Dapps (eth2phone, Airswap, CryptoFighters, ENS name buying) and transcations between Android and iOS users, autotests re-run against latest build https://github.com/status-im/status-react/pull/6877#issuecomment-445738064, confirmed we are good to go!
Ready for merge.

**Comment by cammellos** [comment]
@mandrigin I will squash and merge
